File: specialized/sudoku.nim
#!./make.py --10
#INFO: algo with specialized types using bitsets (optimized by copilot)

import std/[rdstdin, strutils]

# NumSet: represent 1-9 values using bits 0-8 in a uint16
type NumSet = uint16

const ALL_SET: NumSet = 0b1_1111_1111
const EMPTY_SET: NumSet = 0

func oneHot(val: uint8): NumSet =
    NumSet(1) shl int(val)

func popcount(x: NumSet): int =
    # Naive popcount for small sets
    var count = 0
    var bits = x
    while bits != 0:
        count += int(bits and 1)
        bits = bits shr 1
    count

func val(x: NumSet): uint8 =
    # Find the position of the lowest set bit
    for i in 0..<9:
        if (x and (NumSet(1) shl i)) != 0:
            return uint8(i)
    0

# Grid and solver
type Grid = object
    data: array[81, NumSet]
    spaces: seq[int]

proc parseGrid(s: string): Grid =
    var data: array[81, NumSet]
    var spaces: seq[int] = @[]
    
    for i, c in s:
        case c
        of '1'..'9':
            data[i] = oneHot(uint8(c) - uint8('1'))
        of '.':
            data[i] = EMPTY_SET
            spaces.add(i)
        else:
            discard
    
    Grid(data: data, spaces: spaces)

proc gridToString(g: Grid): string =
    var result = ""
    for x in g.data:
        if x == EMPTY_SET:
            result.add('.')
        else:
            result.add(char(uint8('1') + x.val()))
    result

proc sqr(g: Grid, x: int, y: int): NumSet =
    let x = (x div 3) * 3
    let y = (y div 3) * 3
    let i = y * 9 + x
    g.data[i] or g.data[i+1] or g.data[i+2] or
    g.data[i+9] or g.data[i+10] or g.data[i+11] or
    g.data[i+18] or g.data[i+19] or g.data[i+20]

proc col(g: Grid, x: int): NumSet =
    var result: NumSet = 0
    for y in 0..8:
        result = result or g.data[y*9 + x]
    result

proc row(g: Grid, y: int): NumSet =
    var result: NumSet = 0
    for x in 0..8:
        result = result or g.data[y*9 + x]
    result

proc free(g: Grid, x: int, y: int): NumSet =
    let col = col(g, x)
    let row = row(g, y)
    let sqr = sqr(g, x, y)
    ALL_SET and not (col or row or sqr)

proc resolv(g: var Grid, spaceIdx: int): bool =
    if spaceIdx >= g.spaces.len:
        return true
    
    var ibest = -1
    var cbest = ALL_SET
    var cbestLen = 10
    
    for idx in spaceIdx..<g.spaces.len:
        let i = g.spaces[idx]
        let c = free(g, i mod 9, i div 9)
        
        if c == EMPTY_SET:
            return false
        
        let clen = popcount(c)
        if clen < cbestLen:
            ibest = idx
            cbest = c
            cbestLen = clen
        
        if clen == 1:
            break
    
    if ibest >= 0:
        # Swap spaces[spaceIdx] with spaces[ibest]
        let temp = g.spaces[spaceIdx]
        g.spaces[spaceIdx] = g.spaces[ibest]
        g.spaces[ibest] = temp
        
        let s = g.spaces[spaceIdx]
        
        # Try each possibility
        var bit = cbest
        while bit != 0:
            let candidate = bit and (0 - bit) # isolate lowest set bit
            bit = bit xor candidate
            
            g.data[s] = candidate
            if resolv(g, spaceIdx + 1):
                return true
        
        g.data[s] = EMPTY_SET
        
        # Swap back
        let temp2 = g.spaces[spaceIdx]
        g.spaces[spaceIdx] = g.spaces[ibest]
        g.spaces[ibest] = temp2
        
        return false
    
    return true

###############################################
# Main loop

var line: string
while true:
    let ok = readLineFromStdin("", line)
    if not ok: break
    if line.len > 0:
        var grid = parseGrid(line)
        discard resolv(grid, 0)
        echo gridToString(grid) & " "

File: specialized/sudoku.go
//INFO: algo with arrays (optimized by copilot)
package main

import (
	"bufio"
	"fmt"
	"os"
)

// count number of set bits - using manual bit manipulation
func popcount(x uint16) int {
	x -= (x >> 1) & 0x5555
	x = (x & 0x3333) + ((x >> 2) & 0x3333)
	x = (x + (x >> 4)) & 0x0F0F
	return int((x * 0x0101) >> 8)
}

// fast path for getting free numbers using bitmask
func freeMask(g []byte, pos int) uint16 {
	var used uint16
	
	x := pos % 9
	y := pos / 9
	
	// Check row - unrolled
	row_start := y * 9
	if c := g[row_start]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[row_start+1]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[row_start+2]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[row_start+3]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[row_start+4]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[row_start+5]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[row_start+6]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[row_start+7]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[row_start+8]; c != '.' {
		used |= 1 << (c - '0')
	}
	
	// Check column - unrolled
	if c := g[x]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[9+x]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[18+x]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[27+x]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[36+x]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[45+x]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[54+x]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[63+x]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[72+x]; c != '.' {
		used |= 1 << (c - '0')
	}
	
	// Check 3x3 box
	box_x := (x / 3) * 3
	box_y := (y / 3) * 3
	b0 := (box_y * 9) + box_x
	if c := g[b0]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[b0+1]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[b0+2]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[b0+9]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[b0+10]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[b0+11]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[b0+18]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[b0+19]; c != '.' {
		used |= 1 << (c - '0')
	}
	if c := g[b0+20]; c != '.' {
		used |= 1 << (c - '0')
	}
	
	return ^used & 0x3FE // invert and mask bits 1-9
}

// resolv solves the Sudoku grid using backtracking
func resolv(g []byte) bool {
	// Find cell with minimum possibilities
	ibest := -1
	var cbest uint16
	minCount := 10

	for i := 0; i < 81; i++ {
		if g[i] == '.' {
			mask := freeMask(g, i)
			if mask == 0 {
				return false
			}
			cnt := popcount(mask)
			if cnt < minCount {
				minCount = cnt
				ibest = i
				cbest = mask
				if cnt == 1 {
					break
				}
			}
		}
	}

	if ibest >= 0 {
		// Try each possibility
		if (cbest & (1 << 1)) != 0 {
			g[ibest] = '1'
			if resolv(g) {
				return true
			}
			g[ibest] = '.'
		}
		if (cbest & (1 << 2)) != 0 {
			g[ibest] = '2'
			if resolv(g) {
				return true
			}
			g[ibest] = '.'
		}
		if (cbest & (1 << 3)) != 0 {
			g[ibest] = '3'
			if resolv(g) {
				return true
			}
			g[ibest] = '.'
		}
		if (cbest & (1 << 4)) != 0 {
			g[ibest] = '4'
			if resolv(g) {
				return true
			}
			g[ibest] = '.'
		}
		if (cbest & (1 << 5)) != 0 {
			g[ibest] = '5'
			if resolv(g) {
				return true
			}
			g[ibest] = '.'
		}
		if (cbest & (1 << 6)) != 0 {
			g[ibest] = '6'
			if resolv(g) {
				return true
			}
			g[ibest] = '.'
		}
		if (cbest & (1 << 7)) != 0 {
			g[ibest] = '7'
			if resolv(g) {
				return true
			}
			g[ibest] = '.'
		}
		if (cbest & (1 << 8)) != 0 {
			g[ibest] = '8'
			if resolv(g) {
				return true
			}
			g[ibest] = '.'
		}
		if (cbest & (1 << 9)) != 0 {
			g[ibest] = '9'
			if resolv(g) {
				return true
			}
			g[ibest] = '.'
		}
		return false
	}
	return true
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		input := scanner.Bytes()
		if len(input) != 81 {
			continue
		}
		
		g := make([]byte, 81)
		copy(g, input)
		
		if resolv(g) {
			fmt.Println(string(g))
		} else {
			fmt.Println("")
		}
	}
}
